include/zephyr/sys/kref.h

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_KREF_H_
+#define ZEPHYR_INCLUDE_SYS_KREF_H_
+
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup kref_apis Reference count APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief An atomic reference count.
+ *
+ * The main use case is to share the same data between several users and free
+ * it when the count drops to 0. This can also be used on SMP systems.
+ */
+struct k_ref {
+	atomic_t refs;
+};
+
+/** @brief Initialize k_ref at compile-time. */
+#define K_REF_INIT()                                                                               \
+	{                                                                                          \
+		.refs = ATOMIC_INIT(1),                                                            \
+	}
+
+/** @brief Initialize k_ref at runtime. */
+void k_ref_init(struct k_ref *ref);
+
+/**
+ * @brief Increase the reference count.
+ *
+ * @return `ref` is successful or
+ *         NULL if the reference count was 0 already.
```

Comment:
This seems like a confusing API to me.  Do any other libraries work like that?  Generally "take a reference" is understood to be a straightforward/hairless operation.  Why does a caller care whether or not the object was already reference.  "Is this in use by someone else?" is an **extremely** racy question to be asking in parallel code!

Reply:
I copied most of this from bt_conn_ref and bt_conn_unref. It's true that this is somewhat unusual but I always tried to explain this to myself with "All references having the same pointer is an implementation detail that may or may not be true".

Reply:
I removed that handling and added assertions instead.